text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Shrinks the block of memory associated to a handle. On success, the handle is invalidated.
+    unsafe fn shrink(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError>;
+
+    /// Allocates a new handle to a block of zeroed memory. On success, invalidates any existing handle.
+    fn allocate_zeroed(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+
+    /// Grows the block of memory associated to a handle with zeroed memory. On success, the handle is invalidated.
+    unsafe fn grow_zeroed(
+        &self,
+        handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        ...
+    }
+}
+```
+
+_Note:  full-featured documentation of the trait and methods can be found in the companion repository at_
+        https://github.com/matthieu-m/store/blob/main/src/interface.rs.
+
+The `Store` trait is supplemented by 3 additional marker traits, providing extra guarantees:
+
+```rust
+/// A refinement of `Store` which does not invalidate existing handles on allocation, and does not invalidate
+/// unrelated existing handles on deallocation.
+pub unsafe trait MultipleStore: Store {}
+
+/// A refinement of `Store` which does not invalidate existing pointers on allocation, resolution, or deallocation, but
+/// may invalidate them on moves.
+pun unsafe trait StableStore: Store {}
+
+/// A refinement of `Store` which does not invalidate existing pointers, not even on moves. That is, this refinement
+/// guarantees that the blocks of memory are pinned in memory.
+pub unsafe trait PinningStore: StableStore {}
+```
+
+
+##  Library Organization
+
+This RFC proposes to follow the lead of the `Allocator` trait, and add the `Store` traits to the `core` crate, either in
+the `alloc` module or in a new `store` module.
+
+It leaves to a follow-up RFC the introduction of a `store` or `store-collections` crate which would contain the code of
+the various standard collections: `Box`, `Vec`, `String`, `BinaryHeap`, `BTreeMap`, `BTreeSet`, `LinkedList`, and
+`VecDeque`, all adapted for the `Store` API.
+
+Those types would then be re-exported as-is in the `alloc` crate, drastically reducing its size.
+
+
+#   Drawbacks
+
+This RFC increases the surface of the standard library, with yet another `Allocator`.
+
+Furthermore, the natural consequence of adopting this RFC would be rewriting the existing collections in terms of
+`Store`, rather than `Allocator`. A mostly mechanical task, certainly, but an opportunity to introduce subtle bugs in
+the process, even if MIRI would hopefully catch most such bugs.
+
+Finally, having two allocator-like APIs, `Store` and `Allocator`, means that users will forever wonder which trait they
```

Comment:
Do we need two APIs? Couldn't `Store` just replace `Allocator`?

Reply:
That's a good question.

From an "absolute" standpoint the `Store` API can do anything that the `Allocator` API can do, and more, as it's strictly more generic. And thus we could indeed jettison the `Allocator` API. **But**.

Firstly, I think it best to keep this RFC focused on the `Store` API, and not get distracted by peripheral questions. I would defer any question about removing the `Allocator` API to _after_ sufficient experience has been gained by multiple people other than me with the `Store` API to have qualified opinions on the topic.

In particular, even though `Store` can do more, it is more verbose, and possibly more fiddly. I could see `Allocator` being a higher-level API, for those who care not about the intricacies of `Store`, though I'd encourage collections to be written against `Store` so they can be used with both allocators and stores.

Secondly, even though `Allocator` is unstable, it was widely desired and has been there for years, hence I expect that in practice it is used throughout projects and that as such it would be fairly disruptive to outright remove it without a deprecation period. Hence, even if a decision to remove `Allocator` is ultimately reached, between the introduction of `Store` and the end of the deprecation period of `Allocator` the standard library would just have the two competing APIs. Given that the deprecation period for the layout of `Ipv4Addr` took 2 years, the two APIs may coexist for quite a while, and users will have to choose between them in the meantime.

Reply:
I would not expect that many users will implement their own `Allocator`/`Store`, so having another API that slightly simplifies the implementation doesn't seem worth it to me, but I'm not sure.

About the second point, I agree that `Allocator` should have a deprecation period. But I do feel pretty strongly that "there will be two similar _unstable_ APIs for a while where people will have to choose is not a significant downside at all, not worth considering at all imo.
Also, `Allocator` is quite different from the `Ipv4Addr` layout thing because there, people _accidentally_ relied on it by having the offending crates in their dep tree somewhere, for `Allocator`, people have to _actively opt in to instability_.

Reply:
I think that at least _some_ conversation needs to happen with the allocator WG if it hasn't already, since they are planning to stabilize some portion of the `Allocator` feature soon (see https://github.com/rust-lang/rust/issues/32838). At a minimum, it would be good to make sure that nothing stabilized conflicts with the suggestions here.

Reply:
@tgross35 The team members seem aware of this RFC. @Amanieu's [comment](https://github.com/rust-lang/rust/issues/32838#issuecomment-1554883570) proposal to stabilize the `Allocator` trait references the pre-RFC, and @thomcc has been commenting here (bringing up the very issues he's been polishing on `Allocator`).

The `Store` RFC has been crafted to be compatible with `Allocator`, and will follow `Allocator`'s lead, so I don't foresee any compatibility issue.